packages/dds/tree/docs/wip/schema-evolution.md

```diff
@@ -0,0 +1,583 @@
+
+### Restricting Data with a Schema
+
+The example tree diagram above describes what data is in the tree, but it does not describe what data _could possibly_ be in the tree.
+This is the job of a _schema_ - to constrain the shape and content of the tree to a more specific domain.
+
+For example, in the tree above, we can see that the field under `field key D` of `node 2` contains exactly one node - `node 7`.
+Without a schema, we don't know if that's because there is always a single node in that field, or if it's sometimes empty, or sometimes has more than one node.
+This makes it cumbersome to read data from the tree (the API treats every field as a sequence, even if e.g. it's known that there ought to be exactly one node there), and dangerous to write data to the tree (e.g. if this field is supposed to always have exactly one node, what happens when it is deleted?).
+
+A schema could tell us that `node 2` is a "Dog" and that every dog has a string data payload - the dog's name - under `field key D`.
+In addition to knowing that the data in `node 7` is a string, and what that string means, we also know that it wouldn't make sense to put more than one node in that field - a dog can't have two names (not in this domain, anyway).
+
+If we have the guarantee that the data in the tree conforms to this "Dog" schema, then users can be confident about how to read and write the dog's name.
+We can also make the APIs better (for example, having a specific `setName(nameString)` function rather than a general `setField(fieldKey, nodes)` function).
+
+### Tree Schema Structure
+
+What does a tree schema actually look like?
+What aspects of the Intentional Tree Model introduced above might be constrained by a schema?
+The following list describes the building blocks for schema enforcement that are in use by SharedTree and have proven to be sufficient for typical SharedTree projects so far.
+
+1. **Nodes have a _type_.**
+   Knowing what a given node represents in an application domain is important to understanding the data in and around that node.
+   By assigning each node a unique type identifier (e.g. "Dog" or "com.microsoft.powerpoint.rectangle"), the data held by that node can be understood.
+
+2. **Nodes have expected fields.**
+   For a given type of node, a schema enumerates which fields are expected to be under that node.
+   For example, a node of type "Dog" is expected to have a field under the field key "name".
+
+   > Having an expected field means that the field is expected to be there, but the inverse is not true.
+     _Not_ having an expected field with key _K_ does not mean that there _can't_ be a field with key _K_.
+     Field _K_ might be in the tree, but it's invisible to a schema that doesn't know about it and that schema can (in theory) pretend it doesn't exist.
```

Comment:
I wonder how reader should feel about this uncertainty. Is this a statement of current possibilities (ability to ignore extra fields), or is this a statement of possible future changes in capabilities or behaviors?

Reply:
@noencke agree that this can be unclear. May be good to give a concrete example.